### PR TITLE
Support types other than String for partition columns on ListingTables

### LINKED ISSUE #4218: support scan non-string columns partitioned parquet files
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
reference -- delta-io/delta-rs#937

The problem is because partitioned columns in the schema of deltalake dismatches that of datafusion which is always the value type(Utf8) of `DEFAULT_PARTITION_COLUMN_DATATYPE`.

### PR DESCRIPTION
**Pull request by doki23**
# Which issue does this PR close?
Closes #4218.

# Rationale for this change
For supporting more types of partitioned columns.

# What changes are included in this PR?
This pr makes datafusion to extract data type of partitioned columns from file groups instead of setting a default `Utf8` type.

# Are these changes tested?
Already passed all tests including every *_with_partitions test.

# Are there any user-facing changes?
Yes. Users can query partitioned columns by delta table definition.

### CODE REVIEW

**Comment by mingmwang** [comment]
I think not all the types of columns can be used as partition columns, should there a white list for supported types?

**Comment by mingmwang** [comment]
And can you explain a little how you will infer the partition column types or partition spec from the file paths ?
And what happened if there are conflicts?

**Comment by alamb** [comment]
Thank you @doki23  -- I think the functionality needs some tests before we would consider merging it.

**Comment by doki23** [comment]
> I think not all the types of columns can be used as partition columns, should there a white list for supported types?

> And can you explain a little how you will infer the partition column types or partition spec from the file paths ? And what happened if there are conflicts?

Thanks for your reply. I believe the answers are same -- these types are provided by every datasource itself if I am not mistaken. So what I do is to make them matches to each other by their definitions.

**Comment by doki23** [comment]
> Thank you @doki23 -- I think the functionality needs some test of a non-string partition column before we would consider merging it.
> 
> Otherwise how would we know if we caused a regression (aka broke) this feature?

Thanks you. I did test them in delta lake but I forgot to add some unit tests for datafusion, I'll do it.

**Comment by doki23** [comment]
I find that [ListingTable](https://github.com/apache/arrow-datafusion/blob/406c1087bc16f8d2a49e5a9b05d2a0e1b67f7aa5/datafusion/core/src/datasource/listing/table.rs#L416) also treats type of partitioned columns as `Utf8`. I plan to change the struct [ListOptions](https://github.com/apache/arrow-datafusion/blob/406c1087bc16f8d2a49e5a9b05d2a0e1b67f7aa5/datafusion/core/src/datasource/listing/table.rs#L207).
There 2 ways:
1. change the existing property `table_partition_cols` but it's a break change to proto.
2. add an optional property `table_partition_fields` whose type is `Option<Vec<Field>>`.
I am appreciative if any suggestions.

**Comment by doki23** [comment]
@mingmwang @alamb Would you please review my pr? I think it's ready now.

File: datafusion/core/src/datasource/listing_table_factory.rs
Comment (alamb) [inline datafusion/core/src/datasource/listing_table_factory.rs:98]:
```suggestion
                    .map(|x| (x.clone(), DataType::Utf8))
```

File: datafusion/core/src/datasource/listing_table_factory.rs
Comment (alamb) [inline datafusion/core/src/datasource/listing_table_factory.rs:109]: Are you sure it is ok to `unwrap()` here (which will panic)? Given the function returns a `Result` perhaps we could return an error if the field was not present in the schema.

File: datafusion/core/tests/path_partition.rs
Comment (alamb) [inline datafusion/core/tests/path_partition.rs:61]: Nice

File: datafusion/core/src/datasource/listing/table.rs
Comment (alamb) [inline datafusion/core/src/datasource/listing/table.rs:283]: It would help me read the code if this was a named struct so the code could refer to `.name` and `.datatype` rather than `.0` and `.1`. But I don't think it is necessary to merge this PR

```rust
#[derive(Clone)]
struct PartitionColumn {
  name: String,
  data_type: DataType
}
```

File: datafusion/proto/src/logical_plan.rs
Comment (alamb) [inline datafusion/proto/src/logical_plan.rs:893]: 👍

**Review by alamb — APPROVED** [review]
Thanks @doki23  -- this looks good to me. I had some style suggestions but nothing critical.

I'll plan to merge this PR tomorrow -- let me know if you want to delay so you can make any more changes

File: datafusion/core/src/datasource/listing/table.rs
Comment (alamb) [inline datafusion/core/src/datasource/listing/table.rs:550]: I think this is going to `panic` if the user specifies a partition column that is not present.

It would be nice to make it an error -- can you either do so as a follow on PR or file a ticket?

**Review by alamb — APPROVED** [review]
Looks good -- thank you @doki23

**Pull request merged**
